Dispatch to Python via __torch_dispatch__

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59760 Dispatch to Python via __torch_dispatch__**

See https://github.com/pytorch/pytorch/issues/59049

There are some moving parts to this PR, I'll structure this explanation so the straightforward parts go first, and then the less straightforward parts.

**The actual dispatch to Python.** The core logic of dispatch to Python lives in `concrete_dispatch_fn` in `torch/csrc/autograd/python_variable.cpp`. It takes the input IValue stack, scans all the arguments for Tensor arguments, and defers most of the heavy lifting to `handle_torch_function_no_python_arg_parser` which actually does all of the logic for calling out to torch dispatch (in particular, this function handles multiple dispatch situations for you). Because we have a different function name than regular `__torch_function__` handling, `handle_torch_function_no_python_arg_parser` is generalized to accept a magic method name to look for when testing if Tensors have custom handling or not. Unlike `__torch_function__`, by default there is no `__torch_dispatch__` on Tensor classes. 

**Maintaining the Python dispatch key.** In order to get to the dispatch to Python logic, we must tag Tensors with the `__torch_dispatch__` magic method with the newly added Python dispatch key (separated from PythonFuncTorch to allow for a transitional period while they migrate to this mechanism). We expose a new private property `_is_python_dispatch` that assists in debugging if a Tensor is participating in Python dispatch or not. We apply the Python dispatch key the first time a PyObject for a Tensor is constructed (THPVariable_NewWithVar), testing if `__torch_dispatch__` exists with  then newly added `check_has_torch_dispatch`. 

**Shallow copy and detach.** For the simple examples tested in this PR, most creations of Tensor route through the dispatcher. The exception to this is `shallow_copy_and_detach`, which bypasses the dispatcher and is used when saving tensors for backwards. When a Tensor is Python dispatch, we override the behavior of `shallow_copy_and_detach` to instead directly call into `__torch_dispatch__` to perform a `detach` operation (in the same way it would be invoked if you called `detach` directly). Because this Python call is triggered directly from c10::TensorImpl, it must be indirected through `PyInterpreter::detach`, which is the general mechanism for dynamic dispatching to the Python interpreter associated with a TensorImpl.

**torchdeploy compatibility.** The dispatch to Python logic cannot be directly registered to the dispatcher as it is compiled in the Python library, which will get loaded multiple times per torchdeploy interpreter. Thus, we must employ a two phase process. First, we register a fallback inside a non-Python library (aten/src/ATen/core/PythonFallbackKernel.cpp). Its job is to determine the appropriate PyInterpreter to handle the Python dispatch by going through all of the arguments and finding the first argument that has a PyObject/PyInterpreter. With this PyInterpreter, it makes another dynamic dispatch via "dispatch" which will go to the correct torchdeploy interpreter to handle dispatching to actual Python.

**Interaction with `__torch_function__`.** I slightly modified the default `Tensor.__torch_function__` to test if the returned tensor is already at the desired subclass before rewrapping. This makes it interoperate smoothly with `__torch_dispatch__, without requiring users to explicitly disable default `__torch_function__`

**Testing.** We provide a simple example of a LoggingTensor for testing, which can be used to generate TorchScript-like traces to observe what operations are being called when a Tensor is invoked. Although a LoggingTensor would be better implemented via an is-a relationship rather than a has-a relationship (as is done in the test), we've done it this way to show that arbitrarily complex compositions of tensors inside a tensor work properly.

**Known limitations.**

* We haven't adjusted any operator code, so some patterns may not work (as they lose the Python subclass in an unrecoverable way)
* We don't ever populate kwargs, even when an argument is kwarg-only
* Remove arguments when they match the defaults (and provide a different mechanism for getting full defaults)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D29017912](https://our.internmc.facebook.com/intern/diff/D29017912)

ezyang committed:
Dispatch to Python via __torch_dispatch__

See https://github.com/pytorch/pytorch/issues/59049

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Dispatch to Python via __torch_dispatch__"


See https://github.com/pytorch/pytorch/issues/59049

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Dispatch to Python via __torch_dispatch__"


See https://github.com/pytorch/pytorch/issues/59049

There are some moving parts to this PR, I'll structure this explanation so the straightforward parts go first, and then the less straightforward parts.

**The actual dispatch to Python.** The core logic of dispatch to Python lives in `concrete_dispatch_fn` in `torch/csrc/autograd/python_variable.cpp`. It takes the input IValue stack, scans all the arguments for Tensor arguments, and defers most of the heavy lifting to `handle_torch_function_no_python_arg_parser` which actually does all of the logic for calling out to torch dispatch (in particular, this function handles multiple dispatch situations for you). Because we have a different function name than regular `__torch_function__` handling, `handle_torch_function_no_python_arg_parser` is generalized to accept a magic method name to look for when testing if Tensors have custom handling or not. Unlike `__torch_function__`, by default there is no `__torch_dispatch__` on Tensor classes. 

**Maintaining the Python dispatch key.** In order to get to the dispatch to Python logic, we must tag Tensors with the `__torch_dispatch__` magic method with the newly added Python dispatch key (separated from PythonFuncTorch to allow for a transitional period while they migrate to this mechanism). We expose a new private property `_is_python_dispatch` that assists in debugging if a Tensor is participating in Python dispatch or not. We apply the Python dispatch key the first time a PyObject for a Tensor is constructed (THPVariable_NewWithVar), testing if `__torch_dispatch__` exists with  then newly added `check_has_torch_dispatch`. 

**Shallow copy and detach.** For the simple examples tested in this PR, most creations of Tensor route through the dispatcher. The exception to this is `shallow_copy_and_detach`, which bypasses the dispatcher and is used when saving tensors for backwards. When a Tensor is Python dispatch, we override the behavior of `shallow_copy_and_detach` to instead directly call into `__torch_dispatch__` to perform a `detach` operation (in the same way it would be invoked if you called `detach` directly). Because this Python call is triggered directly from c10::TensorImpl, it must be indirected through `PyInterpreter::detach`, which is the general mechanism for dynamic dispatching to the Python interpreter associated with a TensorImpl.

**torchdeploy compatibility.** The dispatch to Python logic cannot be directly registered to the dispatcher as it is compiled in the Python library, which will get loaded multiple times per torchdeploy interpreter. Thus, we must employ a two phase process. First, we register a fallback inside a non-Python library (aten/src/ATen/core/PythonFallbackKernel.cpp). Its job is to determine the appropriate PyInterpreter to handle the Python dispatch by going through all of the arguments and finding the first argument that has a PyObject/PyInterpreter. With this PyInterpreter, it makes another dynamic dispatch via "dispatch" which will go to the correct torchdeploy interpreter to handle dispatching to actual Python.

**Testing.** We provide a simple example of a LoggingTensor for testing, which can be used to generate TorchScript-like traces to observe what operations are being called when a Tensor is invoked. Although a LoggingTensor would be better implemented via an is-a relationship rather than a has-a relationship (as is done in the test), we've done it this way to show that arbitrarily complex compositions of tensors inside a tensor work properly.

**Known limitations.**

* We haven't adjusted any operator code, so some patterns may not work (as they lose the Python subclass in an unrecoverable way)
* It's not possible yet to redispatch excluding `__torch_dispatch__`
* `__torch_function__` must be explicitly disabled with `_disabled_torch_function_impl` otherwise things don't work quite correctly (in particular, what is being disabled is default subclass preservation behavior.)
* We don't ever populate kwargs, even when an argument is kwarg-only

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29017912).

zou3519 commented on test/test_python_dispatch.py:
```diff
@@ -0,0 +1,139 @@
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+        def unwrap(e):
+            if isinstance(e, torch.Tensor):
+                return e.elem
+            elif isinstance(e, (list, tuple)):
+                return [unwrap(sub_e) for sub_e in e]
+            else:
+                return e
```
nit: pytrees handle this a bit more succinctly
```
from torch.utils._pytree import tree_map

def unwrap_single(e):
    return e.elem if isinstance(e, torch.Tensor) else e.elem

def wrap_single(e):
    return LoggingTensor(e) if isinstance(e, torch.Tensor) else e

wrap = functools.partial(tree_map, wrap_single)
unwrap = functools.partial(tree_map, unwrap_single)
```

ezyang replied:
oops, forgot this is a thing, pytrees sounds great

zou3519 commented on test/test_python_dispatch.py:
```diff
@@ -0,0 +1,139 @@
+    __torch_function__ = _disabled_torch_function_impl
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
```
Can you go through an example of what exactly the chain of dispatch is when one calls e.g. torch.sin on a LoggingTensor? Something like an updated version of #59049's "How does holder-less dispatch to Python work" section would be helpful (do we still dispatch to `__torch_function__` first, then that enters the dispatcher, and then that comes around to `__torch_dispatch__` ?

ezyang replied:
I'll put this in code somewhere, but just quickly:

1. Call `torch.sin`
2. Attempt `__torch_function__`. In LoggingTensor torch function is disabled so we bypass it entirely
3. Enter dispatcher, wind your way through Autograd
4. Hit Python dispatch key, call `__torch_dispatch__`

zou3519 commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -112,6 +121,15 @@ static const char* VOLATILE_WARNING =
     "volatile was removed and now has no effect. Use "
     "`with torch.no_grad():` instead.";
 
+bool check_has_torch_dispatch(PyObject *obj) {
```
nit: `static bool check_has_torch_dispatch`, unless you wanted to toss this into a header?

zou3519 commented on c10/core/TensorImpl.h:
```diff
@@ -263,6 +277,18 @@ struct C10_API PyInterpreter {
+  // Perform a detach by deferring to the __torch_dispatch__ implementation of
+  // detach, which will also arrange for the PyObject to get copied in this
+  // situation
+  __ubsan_ignore_function__ c10::intrusive_ptr<TensorImpl> detach(const TensorImpl* self) const {
```
In some cases, `shallow_copy_and_detach` is a little more than a detach, it also resets the version counter(?) (see [link](https://github.com/pytorch/pytorch/blob/e7ad82eb2f9856682d95a69114a04655a32c5ad2/torch/csrc/autograd/variable.cpp#L365-L367)). Is this a problem?

ezyang replied:
Oh alright, I guess we need a different operator for this.

ezyang replied:
Yeah, so we could also just manually fix things up after the user implementation returns. TBH, I'm not sure what the right thing to do here.

albanD commented on test/test_python_dispatch.py:
```diff
@@ -0,0 +1,139 @@
+
+    @staticmethod
+    def __new__(cls, elem, *args, **kwargs):
+        r = torch.Tensor._make_subclass(cls, elem.to('meta'), elem.requires_grad)
```
Is the idea here that the "wrapping" Tensor is a meta Tensor so that it doesn't hold any memory. And the actual Tensor is `elem` that actually has memory.

Does that mean that we can use this "trick" to replace any Tensor-like object into a Tensor subclass at no extra cost?

ezyang replied:
Yes.

bdhirsh replied:
I sorta understood that but the clarification was helpful :)

albanD commented on test/test_python_dispatch.py:
```diff
@@ -0,0 +1,139 @@
+            g, = torch.autograd.grad((y,), (x,), (grad_y,))
+
+        self.assertEqual(g.elem, torch.tensor([6.0]))
+        # $3 and $5 comes from nowhere because detach for saved variable not tracked atm
```
Are these numbers still accurate?

Also https://github.com/pytorch/pytorch/pull/59837 is most likely going to conflict with this, so we should be careful not to land these two at the same time cc @Varal7

ezyang replied:
Oops, this is out of date

@Varal7 don't worry about merging before me, these tests are very easy to update.

albanD reviewed (commented):
That is surprisingly simple! Good job!

ezyang commented on aten/src/ATen/core/PythonFallbackKernel.cpp:
```diff
@@ -0,0 +1,35 @@
+void pythonFallback(const c10::OperatorHandle& op, torch::jit::Stack* stack) {
+  const auto& schema = op.schema();
+  const auto num_arguments = schema.arguments().size();
+  for (const auto& ivalue : torch::jit::last(*stack, num_arguments)) {
```
These are in aten/src/ATen/core/stack.h

ezyang commented on c10/core/TensorImpl.cpp:
```diff
@@ -31,9 +31,19 @@ static void noop_decref_fn(const PyInterpreter*, PyObject*) {
 }
 
+static c10::intrusive_ptr<TensorImpl> noop_detach_fn(const PyInterpreter*, const TensorImpl*) {
+  return c10::intrusive_ptr<TensorImpl>();
```
Not sure. Returning a nullptr here is OKish because `shallow_copy_and_detach` tests to see if a nullptr was returned, and if it was, does the "traditional" shallow copy and detach logic. So your code will work about as well as it possibly could given that the PyInterpreter is dead. It's true we could raise an error, maybe that is safer, but smoothing over the system here was pretty easy and predictable so I did it.

ezyang replied:
Let's just raise an error; even if you successfully shallow copy and detach, if you try to run any op on it you are still going to die.

ezyang commented on c10/core/TensorImpl.cpp:
```diff
@@ -459,6 +471,12 @@ c10::AutogradMetaInterface* TensorImpl::autograd_meta() const {
     const c10::VariableVersion& version_counter,
     bool allow_tensor_metadata_change) const {
+  if (key_set_.has(DispatchKey::Python)) {
+    auto r = pyobj_interpreter_.load(std::memory_order_acquire)->detach(this);
```
Python key implies there is an interpreter. However, I have to test for interpreter on all inputs in fallback logic, because you can have a mixed function call with Python-available Tensors and Python-unavailable Tensors (remember we OR together the dispatch key sets to determine dispatch).

albanD replied:
Ho I forgot that case. Sounds good!

ezyang commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -1436,3 +1467,116 @@ bool THPVariable_initModule(PyObject *module)
+  // For now, overloads get coalesced.  Might be easier for users if they get
+  // overload resolution but is more complicated (need to expose separate
+  // functions per overload)
+  py::handle torch_api_function = py::module::import("torch").attr("ops").attr(ns).attr(func_name);
```
I guess we can cache `py::module::import("torch").attr("ops")`

ezyang replied:
I grepped `git grep py::module::import | grep static` and there are no occurrences, for what it's worth. Non leaking caches are a bit worrisome, IMO, because destruction can interact unpredictably with Python interpreter shutdown...

ezyang replied:
Not going to do this.

ezyang commented on test/test_python_dispatch.py:
```diff
@@ -0,0 +1,139 @@
+    __torch_function__ = _disabled_torch_function_impl
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
```
Yeah, if we decide to never pass kwargs (I'm still not 100% clear what we should do here), we should modify the signature here. It's a little annoying because I have to add more conditional swiss cheese.

ezyang commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -1436,3 +1467,116 @@ bool THPVariable_initModule(PyObject *module)
+    py::list outs = py::cast<py::list>(out);
+    TORCH_INTERNAL_ASSERT(outs.size() == op.schema().returns().size());
+    for (unsigned idx = 0; idx < outs.size(); idx++) {
+      torch::jit::push(stack, torch::jit::toTypeInferredIValue(outs[idx]));
```
Hmm, yeah, we need to do that, good catch.

ezyang commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -1436,3 +1467,116 @@ bool THPVariable_initModule(PyObject *module)
+  if (op.schema().returns().size() == 1) {
+    torch::jit::push(stack, torch::jit::toTypeInferredIValue(out));
+  } else {
+    py::list outs = py::cast<py::list>(out);
```
You're right, this code needs to correctly handle both list and tuple return from PYthon.

bdhirsh commented on aten/src/ATen/core/PythonFallbackKernel.cpp:
```diff
@@ -0,0 +1,35 @@
+      }
+    } else if (ivalue.isList()) {
+      for (const auto& nv : ivalue.toListRef()) {
+        if (nv.isTensor()) {
```
not a big deal, but does this code imply that your `List` ivalue might have mixed types, some of which are Tensors and some that aren't? If not, maybe `isTensorList()` is a little clearer.

ezyang replied:
Oh you're right, shouldn't ever have mixed types

ezyang replied:
Done but note that we can't use toTensorList as that will induce refcount bumps

bdhirsh commented on aten/src/ATen/core/PythonFallbackKernel.cpp:
```diff
@@ -0,0 +1,35 @@
+          auto* interpreter = nv.unsafeToTensorImpl()->pyobj_interpreter();
+          if (interpreter) {
+            interpreter->dispatch(op, stack);
+            return;
```
Still reading, but I see that we dispatch using the interpreter of the first valid tensor argument. Is it possible to have (and should we error check) the case where your op is passed two tensors from different interpreters?

ezyang replied:
This case is possible. When it happens, we will subsequently error inside the actual dispatch call because it will attempt to retrieve PyObjects for each of the Tensor inputs, but the interpreter label for some of them will be wrong. Thus, not necessary to test here!

bdhirsh commented on c10/core/TensorImpl.cpp:
```diff
@@ -153,6 +163,8 @@ TensorImpl::TensorImpl(
 
   key_set = key_set | getAutocastRelatedKeySetFromBackend(k);
 
+  key_set = key_set.remove(DispatchKey::Python);
```
might be worth a quick note on why we remove the python key specifically in TensorImpl constructors? I think I got it after looking at python_variable.cpp: Only tensors that are exposed to python should be capable of dispatching to python. E.g. if we call some C++ function that internally creates a view without exposing that view to python, we don't want it to do any python dispatching.

ezyang replied:
```
// [Note: Python key removal]
// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// In most constructors for TensorImpl, you will see Python key is removed from
// the passed in DispatchKeySet.  Why?
//
// INVARIANT: Python dispatch key is set iff PyObject for the Tensor has a
// nontrivial __torch_dispatch__ implementation.
//
// When a fresh TensorImpl is created, there is *no* PyObject (this only gets
// initialized lazily at the first point in time the Tensor passes into Python).
// So we would violate the invariant.
//
// In practice, what will happen shortly afterwards is that the TensorImpl
// will get its PyObject initialized by Tensor._make_subclass; at this point
// the Python dispatch key will be set and all is well.  The point is to delay
// the dispatch key setting until that point.
```

bdhirsh commented:
> The exception to this is shallow_copy_and_detach, which bypasses the dispatcher and is used when saving tensors for backwards

Just to be explicit on why that is: what's the reasoning for not making `shallow_copy_and_detach` a dispatcher-aware op, like everything else? I can imagine two reasons:
(1) perf, since it's called all the time in autograd and we don't want the dispatcher overhead
(2) Up until dispatch-to-python, it's never had to have multiple implementations to dynamically dispatch between. It's backend agnostic, and doesn't have any specific modal-key behavior aside from the new python behavior.

Those seem pretty valid. But the upside to making it dispatcher-aware would be that we wouldn't need to add all of the heavy custom support for `shallow_copy_and_detach` inside of `PyInterpreter`, right?

bdhirsh commented:
Also, not directly related to this PR, but I was digging through the `python_variable.cpp` code to try to understand it better, and I see that `THPVariable_Wrap()` assumes a PyInterpreter status of `MAYBE_UNINITIALIZED` when there's no interpreter on the tensor object, which means we need to a `CES` to set it on initialization: https://github.com/pytorch/pytorch/blob/ab70e1e9848cf470fa97ee50a5de5e1f27d2c22d/torch/csrc/autograd/python_variable.cpp#L175

Are we forced to pessimistically make that assumption, because we don't know if we're running in a multiple-interpreter environment? It looks like we're forced to do the CES in pretty standard pytorch code:
```
>>> import torch
a = torch.ones(2)
NewWithVar. status=MAYBE_UNINITIALIZED
>>> b = torch.ones(2)
NewWithVar. status=MAYBE_UNINITIALIZED
>>> c = a + b
NewWithVar. status=MAYBE_UNINITIALIZED
>>>
```

That seems like a potentially perf improvement opportunity, since it looks like `THPVariable_Wrap()` is used all the time. Are we able to detect when, e.g., torch deploy isn't running, or we only have one interpreter, and throw in a `DEFINITELY_UNINITIALIZED`?

ezyang commented:
> That seems like a potentially perf improvement opportunity, since it looks like THPVariable_Wrap() is used all the time. Are we able to detect when, e.g., torch deploy isn't running, or we only have one interpreter, and throw in a DEFINITELY_UNINITIALIZED?

I agree there is a perf improvement opportunity here. Actually, it's somewhat related to @swolchok's recent https://github.com/pytorch/pytorch/pull/59419 : if we get an ExclusivelyOwned tensor, we can know that it is DEFINITELY_UNINITIALIZED. But an easier fix is to make a special THPVariable_Wrap variant for factory function Python bindings for when we know we just created the tensor.

ezyang commented on c10/core/TensorImpl.cpp:
```diff
@@ -477,6 +495,12 @@ c10::intrusive_ptr<TensorImpl> TensorImpl::shallow_copy_and_detach(
 c10::intrusive_ptr<TensorImpl> TensorImpl::shallow_copy_and_detach(
     c10::VariableVersion&& version_counter,
     bool allow_tensor_metadata_change) const {
+  if (key_set_.has(DispatchKey::Python)) {
```
Given that it's literally just a condition, then a function call inside, I don't think it's worth deduplicating (beyond the general sense that `shallow_copy_and_detach` ought to be deduplicated)

ezyang commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -121,17 +139,21 @@ static const char* VOLATILE_WARNING =
+    v->cdata = MaybeOwned<Variable>::owned(std::move(_var));
+    const auto& var = THPVariable_Unpack(v);
+    var.unsafeGetTensorImpl()->init_pyobj(self_interpreter.get(), obj, status);
+    // Quick short circuits for well known type that definitely doesn't have
```
nope u right, can get rid of this

ezyang commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -1436,3 +1467,116 @@ bool THPVariable_initModule(PyObject *module)
+
+  std::vector<py::handle> overloaded_args;
+  auto args = py::reinterpret_steal<py::object>(PyTuple_New(num_arguments));
+  py::dict kwargs;  // TODO: actually populate kwargs sometimes?
```
Trailing defaults are always passed (otherwise, you wouldn't know how many arguments to pop off the stack). And yes, the current implementation only ever passes things as positional, but I talk a little about whether or not we should fill kwargs at "Should we populate kwargs?" in https://github.com/pytorch/pytorch/issues/59049

ezyang committed:
Update on "Dispatch to Python via __torch_dispatch__"


See https://github.com/pytorch/pytorch/issues/59049

There are some moving parts to this PR, I'll structure this explanation so the straightforward parts go first, and then the less straightforward parts.

**The actual dispatch to Python.** The core logic of dispatch to Python lives in `concrete_dispatch_fn` in `torch/csrc/autograd/python_variable.cpp`. It takes the input IValue stack, scans all the arguments for Tensor arguments, and defers most of the heavy lifting to `handle_torch_function_no_python_arg_parser` which actually does all of the logic for calling out to torch dispatch (in particular, this function handles multiple dispatch situations for you). Because we have a different function name than regular `__torch_function__` handling, `handle_torch_function_no_python_arg_parser` is generalized to accept a magic method name to look for when testing if Tensors have custom handling or not. Unlike `__torch_function__`, by default there is no `__torch_dispatch__` on Tensor classes. 

**Maintaining the Python dispatch key.** In order to get to the dispatch to Python logic, we must tag Tensors with the `__torch_dispatch__` magic method with the newly added Python dispatch key (separated from PythonFuncTorch to allow for a transitional period while they migrate to this mechanism). We expose a new private property `_is_python_dispatch` that assists in debugging if a Tensor is participating in Python dispatch or not. We apply the Python dispatch key the first time a PyObject for a Tensor is constructed (THPVariable_NewWithVar), testing if `__torch_dispatch__` exists with  then newly added `check_has_torch_dispatch`. 

**Shallow copy and detach.** For the simple examples tested in this PR, most creations of Tensor route through the dispatcher. The exception to this is `shallow_copy_and_detach`, which bypasses the dispatcher and is used when saving tensors for backwards. When a Tensor is Python dispatch, we override the behavior of `shallow_copy_and_detach` to instead directly call into `__torch_dispatch__` to perform a `detach` operation (in the same way it would be invoked if you called `detach` directly). Because this Python call is triggered directly from c10::TensorImpl, it must be indirected through `PyInterpreter::detach`, which is the general mechanism for dynamic dispatching to the Python interpreter associated with a TensorImpl.

**torchdeploy compatibility.** The dispatch to Python logic cannot be directly registered to the dispatcher as it is compiled in the Python library, which will get loaded multiple times per torchdeploy interpreter. Thus, we must employ a two phase process. First, we register a fallback inside a non-Python library (aten/src/ATen/core/PythonFallbackKernel.cpp). Its job is to determine the appropriate PyInterpreter to handle the Python dispatch by going through all of the arguments and finding the first argument that has a PyObject/PyInterpreter. With this PyInterpreter, it makes another dynamic dispatch via "dispatch" which will go to the correct torchdeploy interpreter to handle dispatching to actual Python.

**Testing.** We provide a simple example of a LoggingTensor for testing, which can be used to generate TorchScript-like traces to observe what operations are being called when a Tensor is invoked. Although a LoggingTensor would be better implemented via an is-a relationship rather than a has-a relationship (as is done in the test), we've done it this way to show that arbitrarily complex compositions of tensors inside a tensor work properly.

**Known limitations.**

* We haven't adjusted any operator code, so some patterns may not work (as they lose the Python subclass in an unrecoverable way)
* It's not possible yet to redispatch excluding `__torch_dispatch__`
* `__torch_function__` must be explicitly disabled with `_disabled_torch_function_impl` otherwise things don't work quite correctly (in particular, what is being disabled is default subclass preservation behavior.)
* We don't ever populate kwargs, even when an argument is kwarg-only

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D29017912](https://our.internmc.facebook.com/intern/diff/D29017912)

[ghstack-poisoned]

ezyang commented:
> Just to be explicit on why that is: what's the reasoning for not making shallow_copy_and_detach a dispatcher-aware op, like everything else?

@bdhirsh talked about this in 1:1 but I want to record it here for posterity. At the end of the day, there needs to be some code that is actually doing the nuts and bolts movement of data between properties of TensorImpl. It might as well be a method on TensorImpl, because then it will actually have access to protected members (which a dispatcher operator won't have). So the fix is not to move `shallow_copy_and_detach` into native per se; rather, it's to make sure that all calls to this function happen from inside dispatched kernel (so that there is an opportunity to interpose.) This would be a useful activity, but there are a lot of `shallow_copy_and_detach` call sites and I didn't feel like litigating them yet; thus, why it's implemented this way.

ezyang committed:
Update on "Dispatch to Python via __torch_dispatch__"


See https://github.com/pytorch/pytorch/issues/59049

There are some moving parts to this PR, I'll structure this explanation so the straightforward parts go first, and then the less straightforward parts.

**The actual dispatch to Python.** The core logic of dispatch to Python lives in `concrete_dispatch_fn` in `torch/csrc/autograd/python_variable.cpp`. It takes the input IValue stack, scans all the arguments for Tensor arguments, and defers most of the heavy lifting to `handle_torch_function_no_python_arg_parser` which actually does all of the logic for calling out to torch dispatch (in particular, this function handles multiple dispatch situations for you). Because we have a different function name than regular `__torch_function__` handling, `handle_torch_function_no_python_arg_parser` is generalized to accept a magic method name to look for when testing if Tensors have custom handling or not. Unlike `__torch_function__`, by default there is no `__torch_dispatch__` on Tensor classes. 

**Maintaining the Python dispatch key.** In order to get to the dispatch to Python logic, we must tag Tensors with the `__torch_dispatch__` magic method with the newly added Python dispatch key (separated from PythonFuncTorch to allow for a transitional period while they migrate to this mechanism). We expose a new private property `_is_python_dispatch` that assists in debugging if a Tensor is participating in Python dispatch or not. We apply the Python dispatch key the first time a PyObject for a Tensor is constructed (THPVariable_NewWithVar), testing if `__torch_dispatch__` exists with  then newly added `check_has_torch_dispatch`. 

**Shallow copy and detach.** For the simple examples tested in this PR, most creations of Tensor route through the dispatcher. The exception to this is `shallow_copy_and_detach`, which bypasses the dispatcher and is used when saving tensors for backwards. When a Tensor is Python dispatch, we override the behavior of `shallow_copy_and_detach` to instead directly call into `__torch_dispatch__` to perform a `detach` operation (in the same way it would be invoked if you called `detach` directly). Because this Python call is triggered directly from c10::TensorImpl, it must be indirected through `PyInterpreter::detach`, which is the general mechanism for dynamic dispatching to the Python interpreter associated with a TensorImpl.

**torchdeploy compatibility.** The dispatch to Python logic cannot be directly registered to the dispatcher as it is compiled in the Python library, which will get loaded multiple times per torchdeploy interpreter. Thus, we must employ a two phase process. First, we register a fallback inside a non-Python library (aten/src/ATen/core/PythonFallbackKernel.cpp). Its job is to determine the appropriate PyInterpreter to handle the Python dispatch by going through all of the arguments and finding the first argument that has a PyObject/PyInterpreter. With this PyInterpreter, it makes another dynamic dispatch via "dispatch" which will go to the correct torchdeploy interpreter to handle dispatching to actual Python.

**Testing.** We provide a simple example of a LoggingTensor for testing, which can be used to generate TorchScript-like traces to observe what operations are being called when a Tensor is invoked. Although a LoggingTensor would be better implemented via an is-a relationship rather than a has-a relationship (as is done in the test), we've done it this way to show that arbitrarily complex compositions of tensors inside a tensor work properly.

**Known limitations.**

* We haven't adjusted any operator code, so some patterns may not work (as they lose the Python subclass in an unrecoverable way)
* It's not possible yet to redispatch excluding `__torch_dispatch__`
* `__torch_function__` must be explicitly disabled with `_disabled_torch_function_impl` otherwise things don't work quite correctly (in particular, what is being disabled is default subclass preservation behavior.)
* We don't ever populate kwargs, even when an argument is kwarg-only

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D29017912](https://our.internmc.facebook.com/intern/diff/D29017912)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29017912).

ezyang commented:
This is ready for final review!

ezyang committed:
Update on "Dispatch to Python via __torch_dispatch__"


See https://github.com/pytorch/pytorch/issues/59049

There are some moving parts to this PR, I'll structure this explanation so the straightforward parts go first, and then the less straightforward parts.

**The actual dispatch to Python.** The core logic of dispatch to Python lives in `concrete_dispatch_fn` in `torch/csrc/autograd/python_variable.cpp`. It takes the input IValue stack, scans all the arguments for Tensor arguments, and defers most of the heavy lifting to `handle_torch_function_no_python_arg_parser` which actually does all of the logic for calling out to torch dispatch (in particular, this function handles multiple dispatch situations for you). Because we have a different function name than regular `__torch_function__` handling, `handle_torch_function_no_python_arg_parser` is generalized to accept a magic method name to look for when testing if Tensors have custom handling or not. Unlike `__torch_function__`, by default there is no `__torch_dispatch__` on Tensor classes. 

**Maintaining the Python dispatch key.** In order to get to the dispatch to Python logic, we must tag Tensors with the `__torch_dispatch__` magic method with the newly added Python dispatch key (separated from PythonFuncTorch to allow for a transitional period while they migrate to this mechanism). We expose a new private property `_is_python_dispatch` that assists in debugging if a Tensor is participating in Python dispatch or not. We apply the Python dispatch key the first time a PyObject for a Tensor is constructed (THPVariable_NewWithVar), testing if `__torch_dispatch__` exists with  then newly added `check_has_torch_dispatch`. 

**Shallow copy and detach.** For the simple examples tested in this PR, most creations of Tensor route through the dispatcher. The exception to this is `shallow_copy_and_detach`, which bypasses the dispatcher and is used when saving tensors for backwards. When a Tensor is Python dispatch, we override the behavior of `shallow_copy_and_detach` to instead directly call into `__torch_dispatch__` to perform a `detach` operation (in the same way it would be invoked if you called `detach` directly). Because this Python call is triggered directly from c10::TensorImpl, it must be indirected through `PyInterpreter::detach`, which is the general mechanism for dynamic dispatching to the Python interpreter associated with a TensorImpl.

**torchdeploy compatibility.** The dispatch to Python logic cannot be directly registered to the dispatcher as it is compiled in the Python library, which will get loaded multiple times per torchdeploy interpreter. Thus, we must employ a two phase process. First, we register a fallback inside a non-Python library (aten/src/ATen/core/PythonFallbackKernel.cpp). Its job is to determine the appropriate PyInterpreter to handle the Python dispatch by going through all of the arguments and finding the first argument that has a PyObject/PyInterpreter. With this PyInterpreter, it makes another dynamic dispatch via "dispatch" which will go to the correct torchdeploy interpreter to handle dispatching to actual Python.

**Testing.** We provide a simple example of a LoggingTensor for testing, which can be used to generate TorchScript-like traces to observe what operations are being called when a Tensor is invoked. Although a LoggingTensor would be better implemented via an is-a relationship rather than a has-a relationship (as is done in the test), we've done it this way to show that arbitrarily complex compositions of tensors inside a tensor work properly.

**Known limitations.**

* We haven't adjusted any operator code, so some patterns may not work (as they lose the Python subclass in an unrecoverable way)
* It's not possible yet to redispatch excluding `__torch_dispatch__`
* `__torch_function__` must be explicitly disabled with `_disabled_torch_function_impl` otherwise things don't work quite correctly (in particular, what is being disabled is default subclass preservation behavior.)
* We don't ever populate kwargs, even when an argument is kwarg-only

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D29017912](https://our.internmc.facebook.com/intern/diff/D29017912)

[ghstack-poisoned]

ezyang committed:
Update on "Dispatch to Python via __torch_dispatch__"


See https://github.com/pytorch/pytorch/issues/59049

There are some moving parts to this PR, I'll structure this explanation so the straightforward parts go first, and then the less straightforward parts.

**The actual dispatch to Python.** The core logic of dispatch to Python lives in `concrete_dispatch_fn` in `torch/csrc/autograd/python_variable.cpp`. It takes the input IValue stack, scans all the arguments for Tensor arguments, and defers most of the heavy lifting to `handle_torch_function_no_python_arg_parser` which actually does all of the logic for calling out to torch dispatch (in particular, this function handles multiple dispatch situations for you). Because we have a different function name than regular `__torch_function__` handling, `handle_torch_function_no_python_arg_parser` is generalized to accept a magic method name to look for when testing if Tensors have custom handling or not. Unlike `__torch_function__`, by default there is no `__torch_dispatch__` on Tensor classes. 

**Maintaining the Python dispatch key.** In order to get to the dispatch to Python logic, we must tag Tensors with the `__torch_dispatch__` magic method with the newly added Python dispatch key (separated from PythonFuncTorch to allow for a transitional period while they migrate to this mechanism). We expose a new private property `_is_python_dispatch` that assists in debugging if a Tensor is participating in Python dispatch or not. We apply the Python dispatch key the first time a PyObject for a Tensor is constructed (THPVariable_NewWithVar), testing if `__torch_dispatch__` exists with  then newly added `check_has_torch_dispatch`. 

**Shallow copy and detach.** For the simple examples tested in this PR, most creations of Tensor route through the dispatcher. The exception to this is `shallow_copy_and_detach`, which bypasses the dispatcher and is used when saving tensors for backwards. When a Tensor is Python dispatch, we override the behavior of `shallow_copy_and_detach` to instead directly call into `__torch_dispatch__` to perform a `detach` operation (in the same way it would be invoked if you called `detach` directly). Because this Python call is triggered directly from c10::TensorImpl, it must be indirected through `PyInterpreter::detach`, which is the general mechanism for dynamic dispatching to the Python interpreter associated with a TensorImpl.

**torchdeploy compatibility.** The dispatch to Python logic cannot be directly registered to the dispatcher as it is compiled in the Python library, which will get loaded multiple times per torchdeploy interpreter. Thus, we must employ a two phase process. First, we register a fallback inside a non-Python library (aten/src/ATen/core/PythonFallbackKernel.cpp). Its job is to determine the appropriate PyInterpreter to handle the Python dispatch by going through all of the arguments and finding the first argument that has a PyObject/PyInterpreter. With this PyInterpreter, it makes another dynamic dispatch via "dispatch" which will go to the correct torchdeploy interpreter to handle dispatching to actual Python.

**Testing.** We provide a simple example of a LoggingTensor for testing, which can be used to generate TorchScript-like traces to observe what operations are being called when a Tensor is invoked. Although a LoggingTensor would be better implemented via an is-a relationship rather than a has-a relationship (as is done in the test), we've done it this way to show that arbitrarily complex compositions of tensors inside a tensor work properly.

**Known limitations.**

* We haven't adjusted any operator code, so some patterns may not work (as they lose the Python subclass in an unrecoverable way)
* It's not possible yet to redispatch excluding `__torch_dispatch__`
* `__torch_function__` must be explicitly disabled with `_disabled_torch_function_impl` otherwise things don't work quite correctly (in particular, what is being disabled is default subclass preservation behavior.)
* We don't ever populate kwargs, even when an argument is kwarg-only

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D29017912](https://our.internmc.facebook.com/intern/diff/D29017912)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29017912).

ezyang committed:
Update on "Dispatch to Python via __torch_dispatch__"


See https://github.com/pytorch/pytorch/issues/59049

There are some moving parts to this PR, I'll structure this explanation so the straightforward parts go first, and then the less straightforward parts.

**The actual dispatch to Python.** The core logic of dispatch to Python lives in `concrete_dispatch_fn` in `torch/csrc/autograd/python_variable.cpp`. It takes the input IValue stack, scans all the arguments for Tensor arguments, and defers most of the heavy lifting to `handle_torch_function_no_python_arg_parser` which actually does all of the logic for calling out to torch dispatch (in particular, this function handles multiple dispatch situations for you). Because we have a different function name than regular `__torch_function__` handling, `handle_torch_function_no_python_arg_parser` is generalized to accept a magic method name to look for when testing if Tensors have custom handling or not. Unlike `__torch_function__`, by default there is no `__torch_dispatch__` on Tensor classes. 

**Maintaining the Python dispatch key.** In order to get to the dispatch to Python logic, we must tag Tensors with the `__torch_dispatch__` magic method with the newly added Python dispatch key (separated from PythonFuncTorch to allow for a transitional period while they migrate to this mechanism). We expose a new private property `_is_python_dispatch` that assists in debugging if a Tensor is participating in Python dispatch or not. We apply the Python dispatch key the first time a PyObject for a Tensor is constructed (THPVariable_NewWithVar), testing if `__torch_dispatch__` exists with  then newly added `check_has_torch_dispatch`. 

**Shallow copy and detach.** For the simple examples tested in this PR, most creations of Tensor route through the dispatcher. The exception to this is `shallow_copy_and_detach`, which bypasses the dispatcher and is used when saving tensors for backwards. When a Tensor is Python dispatch, we override the behavior of `shallow_copy_and_detach` to instead directly call into `__torch_dispatch__` to perform a `detach` operation (in the same way it would be invoked if you called `detach` directly). Because this Python call is triggered directly from c10::TensorImpl, it must be indirected through `PyInterpreter::detach`, which is the general mechanism for dynamic dispatching to the Python interpreter associated with a TensorImpl.

**torchdeploy compatibility.** The dispatch to Python logic cannot be directly registered to the dispatcher as it is compiled in the Python library, which will get loaded multiple times per torchdeploy interpreter. Thus, we must employ a two phase process. First, we register a fallback inside a non-Python library (aten/src/ATen/core/PythonFallbackKernel.cpp). Its job is to determine the appropriate PyInterpreter to handle the Python dispatch by going through all of the arguments and finding the first argument that has a PyObject/PyInterpreter. With this PyInterpreter, it makes another dynamic dispatch via "dispatch" which will go to the correct torchdeploy interpreter to handle dispatching to actual Python.

**Testing.** We provide a simple example of a LoggingTensor for testing, which can be used to generate TorchScript-like traces to observe what operations are being called when a Tensor is invoked. Although a LoggingTensor would be better implemented via an is-a relationship rather than a has-a relationship (as is done in the test), we've done it this way to show that arbitrarily complex compositions of tensors inside a tensor work properly.

**Known limitations.**

* We haven't adjusted any operator code, so some patterns may not work (as they lose the Python subclass in an unrecoverable way)
* `__torch_function__` must be explicitly disabled with `_disabled_torch_function_impl` otherwise things don't work quite correctly (in particular, what is being disabled is default subclass preservation behavior.)
* We don't ever populate kwargs, even when an argument is kwarg-only

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D29017912](https://our.internmc.facebook.com/intern/diff/D29017912)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29017912).

bdhirsh commented on test/test_python_dispatch.py:
```diff
@@ -0,0 +1,257 @@
+
+# TODO: move this into library proper
+@contextlib.contextmanager
+def no_dispatch() -> Iterator[None]:
```
Just want to confirm two questions around `__torch_dispatch__` best practices, assuming the two common use cases of:
- "is a" (where the subclass uses itself as the actual tensor, and doesn't contain any tensor members)
- "has a" (where the subclass just treats itself as a meta tensor, and has at least one tensor member that it calls ops from `__torch_dispatch__` on)

(a) a context manager like `no_dispatch()` isn't needed in the vanilla "has a" case, and is only important in the "is a" case. Assuming that the member tensors are just ordinary tensors and not themselves subclasses, we don't have to worry about recursively hitting the python key again.

(b) Depending on if/how we decide to expose a redispatching API, it seems like that could probably replace the need for this context manager (at least in the vanilla "is a" case)? Since `DispatchKey::Python` has lower precedence than all other non-backend keys, we'd expect to go through all the other behavioral keys before hitting torch_dispatch, so the common case would probably be to just directly dispatch to the backend kernel.

I.e. instead of
```
with no_dispatch():
            output = func(*args)
```
We have something like
```
output = func.redispatch(*args) # dispatches below Python key, to backend kernel
```

ezyang replied:
Yes, agreed on both counts.

zou3519 commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -112,6 +121,15 @@ static const char* VOLATILE_WARNING =
+  PyTypeObject *tp = Py_TYPE(obj);
+  return (
+    !THPVariable_CheckTypeExact(tp) &&
+    // TODO: test if Python key is disabled
+    PyObject_FastGetAttrString(obj, "__torch_dispatch__").ptr() != nullptr
```
Is this TODO happening now or later?

ezyang replied:
actually I think this TODO is obsolete

zou3519 reviewed (commented):
Got through everything -- the PR body and Ed's torchdeploy podcast were helpful in figuring out exactly what was going on. There should probably be a TODO/followup item for what happens with operations on tensor.data (we discussed offline that tensor.data uses shallow_copy_and_detach to create a Tensor with a new version counter))

ezyang commented on test/test_python_dispatch.py:
```diff
@@ -0,0 +1,257 @@
+# 3. Enter dispatcher, wind your way through Autograd
+# 4. Hit Python dispatch key, call __torch_dispatch__
+
+# TODO: TensorBase should work
```
TensorBase still doesn't work for some autograd API functions where they test `isinstance(a, Tensor)` rather than `isinstance(a, TensorBase)`
